operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/PrimaryUpdateAndCacheUtils.java

```diff
@@ -229,4 +236,152 @@ private static <P extends HasMetadata> P pollLocalCache(
       throw new OperatorException(e);
     }
   }
+
+  /** Adds finalizer using JSON Patch. Retries conflicts and unprocessable content (HTTP 422) */
+  @SuppressWarnings("unchecked")
+  public static <P extends HasMetadata> P addFinalizer(Context<P> context, String finalizer) {
+    return addFinalizer(context.getClient(), context.getPrimaryResource(), finalizer);
+  }
+
+  /** Adds finalizer using JSON Patch. Retries conflicts and unprocessable content (HTTP 422) */
+  @SuppressWarnings("unchecked")
+  public static <P extends HasMetadata> P addFinalizer(
+      KubernetesClient client, P resource, String finalizerName) {
+    return conflictRetryingPatch(
+        client,
+        resource,
+        r -> {
+          r.addFinalizer(finalizerName);
+          return r;
+        },
+        r -> !r.hasFinalizer(finalizerName));
+  }
+
+  public static <P extends HasMetadata> P removeFinalizer(
+      Context<P> context, String finalizerName) {
+    return removeFinalizer(context.getClient(), context.getPrimaryResource(), finalizerName);
+  }
+
+  public static <P extends HasMetadata> P removeFinalizer(
+      KubernetesClient client, P resource, String finalizerName) {
+    return conflictRetryingPatch(
+        client,
+        resource,
+        r -> {
+          r.removeFinalizer(finalizerName);
+          return r;
+        },
+        r -> r.hasFinalizer(finalizerName));
+  }
+
+  /**
+   * @param client KubernetesClient
+   * @param resource to update
+   * @param resourceChangesOperator changes to be done on the resource before update
+   * @param preCondition condition to check if the patch operation still needs to be performed or
+   *     not.
+   * @return updated resource or unchanged if the precondition does not hold.
+   * @param <P> resource type
+   */
+  @SuppressWarnings("unchecked")
+  public static <P extends HasMetadata> P conflictRetryingPatch(
```

Comment:
Please add proper documentation, only the parameters are documented, not the purpose or the subtleties of the logic.

Reply:
added, thank you!